### PR TITLE
refactor STS code

### LINKED ISSUE #29753: refactoring: clean up duplicated STS code
security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go and security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go are solving the same problem and communicating with the same thing (google STS) but all the code is duplicated

### PR DESCRIPTION
**Pull request by kailun-qin**
Currently, we have two STS clients/plugins that are solving the same problem and
communicating with the same thing (google STS) but all the code is duplicated.
Besides, we have abstractions e.g. `TokenProvider`, `TokenExchanger` and
`TokenManager` which may bring a few confusions sometimes.

This patch tries to clean up and refactor STS code a bit to address the above
concerns.

Fixes https://github.com/istio/istio/issues/29753

Signed-off-by: Kailun Qin <kailun.qin@intel.com>

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pilot/cmd/pilot-agent/security/security.go [hunk 1]
```diff
@@ -60,10 +58,6 @@ func SetupSecurityOptions(proxyConfig meshconfig.ProxyConfig, secOpt security.Op
 		log.Infof("using credential fetcher of %s type in %s trust domain", credFetcherTypeEnv, o.TrustDomain)
 		o.CredFetcher = credFetcher
 	}
-	// TODO extract this logic out to a plugin
-	if o.CAProviderName == "GoogleCA" || strings.Contains(o.CAEndpoint, "googleapis.com") {
-		o.TokenExchanger = stsclient.NewSecureTokenServiceExchanger(o.CredFetcher, o.TrustDomain)
```
Comment (lei-tang) [inline pilot/cmd/pilot-agent/security/security.go:65]: Are the functions in TokenExchanger merged into TokenManager?

File: security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go [hunk 2]
```diff
@@ -306,6 +310,38 @@ func (p *Plugin) fetchFederatedToken(parameters security.StsRequestParameters) (
 	return respData, nil
 }
 
+// ExchangeToken exchanges oauth access token from K8s SA jwt.
+func (p *Plugin) ExchangeToken(k8sSAjwt string) (string, error) {
```
Comment (lei-tang) [inline security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go:314]: Is this function replacing "func (p *SecureTokenServiceExchanger) ExchangeToken(k8sSAjwt string)"? In original ExchangeToken(), there is a retry mechanism. Why is the retry mechanism removed?

func (p *SecureTokenServiceExchanger) ExchangeToken(k8sSAjwt string) (string, error) {
         ....
	body, err := p.requestWithRetry(jsonStr)
         ...
}

File: security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient_test.go [hunk 3]
```diff
@@ -1,92 +0,0 @@
...
-	"time"
-
-	"istio.io/istio/pkg/test/util/retry"
-	"istio.io/istio/security/pkg/monitoring"
-	"istio.io/istio/security/pkg/nodeagent/util"
-	"istio.io/istio/security/pkg/stsservice/tokenmanager/google/mock"
-	"istio.io/istio/tests/util/leak"
-)
-
-func TestGetFederatedToken(t *testing.T) {
```
Comment (lei-tang) [inline security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient_test.go:30]: Where is the equivalent of this test in the refactored code?

**Comment by lei-tang** [comment]
The STS code is used in multiple places (e.g., Google CA, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow) and there are subtle differences of its usages in different flows. For example, the Google CA flow directly uses the access token of the federated token whereas other flows call constructGenerateAccessTokenRequest(fResp *federatedTokenResponse) to further exchange the federated token to another access token. 
To reduce the risks of breaking these flows that use STS, the refactoring should have tests to show that these flows are working normally.

**Comment by lei-tang** [comment]
Can you describe what changes this refactoring makes?

**Comment by lei-tang** [comment]
@kailun-qin As STS is crucial to multiple flows (e.g., Google CA flow, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow), can you present the PR and its design on Istio Security Group Meeting (https://docs.google.com/document/d/12Xz8fCuql2tPL-cpdLXGGE44kYT4o9NB6mljSiS_01U/edit#heading=h.o8pz6aqnzzgk)?

**Comment by kailun-qin** [comment]
> @kailun-qin As STS is crucial to multiple flows (e.g., Google CA flow, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow), can you present the PR and its design on Istio Security Group Meeting (https://docs.google.com/document/d/12Xz8fCuql2tPL-cpdLXGGE44kYT4o9NB6mljSiS_01U/edit#heading=h.o8pz6aqnzzgk)?

@lei-tang 
Excuse me for the late reply (just busy with some internal stuff...) and thanks a lot for all the comments!
Sure, let me go through the questions and discuss in the next Security WG meeting.

**Comment by lei-tang** [comment]
> > @kailun-qin As STS is crucial to multiple flows (e.g., Google CA flow, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow), can you present the PR and its design on Istio Security Group Meeting (https://docs.google.com/document/d/12Xz8fCuql2tPL-cpdLXGGE44kYT4o9NB6mljSiS_01U/edit#heading=h.o8pz6aqnzzgk)?
> 
> @lei-tang
> Excuse me for the late reply (just busy with some internal stuff...) and thanks a lot for all the comments!
> Sure, let me go through the questions and discuss in the next Security WG meeting.

Thank you! Can you share a design document for review?

**Comment by kailun-qin** [comment]
> > > @kailun-qin As STS is crucial to multiple flows (e.g., Google CA flow, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow), can you present the PR and its design on Istio Security Group Meeting (https://docs.google.com/document/d/12Xz8fCuql2tPL-cpdLXGGE44kYT4o9NB6mljSiS_01U/edit#heading=h.o8pz6aqnzzgk)?
> > 
> > 
> > @lei-tang
> > Excuse me for the late reply (just busy with some internal stuff...) and thanks a lot for all the comments!
> > Sure, let me go through the questions and discuss in the next Security WG meeting.
> 
> Thank you! Can you share a design document for review?

Sure, let me prepare for it. I’ll keep you updated once it’s ready.

**Pull request closed without merging**